File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    int dfs(TreeNode* root, int &maxi){
        if(root==NULL)
        return 0;


        int lhs = dfs(root->left,maxi);
        int rhs = dfs(root->right,maxi);
        
        maxi = max(maxi,lhs+rhs);
        return 1 + max(lhs,rhs);
    }


    int diameterOfBinaryTree(TreeNode* root) {
        // return the length of longest path between any two nodes
        
        int maxi = 0;

        dfs(root,maxi);

        return maxi;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        # find the diameter of binary tree 
        self.maxi = -inf
        def func(root,maxi):
            if not root:
                return 0

            
            lh = func(root.left,maxi)
            rh = func(root.right,maxi)

            self.maxi = max(maxi, lh+rh)

            return 1 + max(lh,rh)

        
        func(root,self.maxi)

        return  self.maxi 


